src/libcrun/kontain_nonkmexec.c

```diff
@@ -0,0 +1,343 @@
+ * We need to allow non kontain programs to be run inside of a kontain container.
+ * But, we don't want arbitrary programs to be run.  So, we have a list of acceptable
+ * programs that "docker exec ...." should be able to run.  This list is stored in
+ * file (see KONTAIN_KRUN_CONFIG definition below).  When it is time to do the
+ * exec() system call we parse the file, check to see if the program we are execing
+ * too is allowed to run without km.  If it is allowed we exec the program directly.
+ * If not, the program will need to run within a km virtual machine.
+ * This blob of code is just used to see if a program should be allowed to run without
+ * km containment.  The caller of libcrun_kontain_nonkmexec_allowed() uses the returned
+ * information to form the appropriate exec arguments.
+ */
+
+struct execpath_entry {
+   SLIST_ENTRY(execpath_entry) link;
+   regex_t re;
+   char* path;
+   char sha256[SHA256_BLOCK_SIZE * 2 + 1];
+};
+typedef struct execpath_entry execpath_entry_t;
+
+SLIST_HEAD(execpath_head, execpath_entry);
+typedef struct execpath_head execpath_head_t;
+
+struct execpath_state {
+   pthread_mutex_t execpath_mutex;          // do we need this?
+   execpath_head_t execpath_head;
+   char* execpath_dbpath;                   // the path we read to populate the execpath_head
+   struct timespec execpath_mtime;          // dbpath mtime when the file was read
+};
+typedef struct execpath_state execpath_state_t;
+
+#define KONTAIN_KRUN_CONFIG	"/var/lib/krun/config"
+
+execpath_state_t execpath_state = {
+   PTHREAD_MUTEX_INITIALIZER,
+   SLIST_HEAD_INITIALIZER(&execpath_state.execpath_head),
+   KONTAIN_KRUN_CONFIG,
+   { 0, 0 }
+};
+
+static void
+kontain_execpath_freedb(execpath_state_t* statep)
+{
+   execpath_entry_t* epp;
+
+   while (!SLIST_EMPTY(&statep->execpath_head)) {
+      epp = SLIST_FIRST(&statep->execpath_head);
+      SLIST_REMOVE_HEAD(&statep->execpath_head, link);
+      free(epp->path);
+      regfree(&epp->re);
+      free(epp);
+   }
+   memset(&statep->execpath_mtime, 0, sizeof(statep->execpath_mtime));
+}
+
+// Destroy the execpath ok db.
+void
+libcrun_kontain_nonkmexec_clean(void)
+{
+   kontain_execpath_freedb(&execpath_state);
+}
+
+static void
+hash_2_ascii(BYTE hash[SHA256_BLOCK_SIZE], char *ascii)
+{
+   int i;
+   for (i = 0; i < SHA256_BLOCK_SIZE; i++) {
+      snprintf(&ascii[i * 2], 3, "%02x", hash[i]);
+   }
+   ascii[i * 2] = 0;
+}
+
+/*
+ * Compute the sha256 hash of the passed file.
+ * Return the hash in ascii in returned_hash.
+ * Returns:
+ *   0 - success
+ *   != 0 - an errno value describing what failed
+ */
+static int
+sha256_file(char* file, char *returned_hash)
+{
+   int f;
+   unsigned char filebuf[128*1024];
+   SHA256_CTX ctx;
+   BYTE final_hash[SHA256_BLOCK_SIZE];
+   ssize_t bytesread;
+
+   f = open(file, O_RDONLY);
+   if (f < 0) {
+      return errno;
+   }
+
+   sha256_init(&ctx);
+   while ((bytesread = read(f, filebuf, sizeof(filebuf))) > 0) {
```

Comment:
would not it be easier to simply mmap the file and then scan memory ?

Reply:
It might be.
The mmap interface is way more complicated than open read close.
If performance is a problem here and mmap helps we can switch.